File: new/twosum_using_hashmap_one_pass.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // create a hashmap of previously visited values and indices. 
        // if the difference between target and current value is in the previously visited hashmap return 
        // the indices, if not return empty list. 
        unordered_map<int, int> previouslyVisited;
        for (int i = 0; i < nums.size(); i++) {
            int difference = target - nums[i];
            if (previouslyVisited.count(difference)) {
                return {previouslyVisited[difference], i};
            }
            previouslyVisited[nums[i]] = i;
        }

        return {};
    }

    // time: O(n)
    // space: O(n)
};

File: new/python/twosum_using_hashmap_one_pass.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # using hashmap one pass
        # create an empty hashmap called prev, where will will put previous visited values
        # if when finding out difference, if the difference is one of the previous visited values in hash map
        # then return those indices
        previously_visited = {}
        for i, val in enumerate(nums):
            difference = target - nums[i]
            if difference in previously_visited:
                return [previously_visited[difference], i]
            previously_visited[val] = i
        return []
        # time: O(n)
        # space: O(n)
